src/instruments/src/ATC/index.tsx

```diff
@@ -0,0 +1,132 @@
+import './style.scss';
+import React, { useEffect, useState } from 'react';
+import { render } from '../Common';
+import { useSimVar, useSplitSimVar } from '../Common/simVars';
+import { useInteractionEvent, useUpdate } from '../Common/hooks';
+
+const getDigitsFromCode = (code: number): number[] => {
+    let codeCopy = code;
+    const digits: number[] = [];
+    while (codeCopy > 0) {
+        digits.push(codeCopy % 10);
+        codeCopy = Math.floor(codeCopy / 10);
+    }
+    digits.reverse();
+    return digits;
+};
+const getBco16FromDigits = (digits: number[]): number => (digits[0] * 4096) + (digits[1] * 256) + (digits[2] * 16) + digits[3];
+
+const PoweredXpdrDisplay = () => {
+    const [currentDigits, setCurrentDigits] = useState([0, 0, 0, 0]);
+    const [currentEnteredDigits, setEnteredDigits] = useState(4);
```

Comment:
So with this component I can already see the issue already on the first line.

Ultimately you're duplicating some kind of source of truth (the transponder sim var) into local state. This is not required, and is generally annoying because now suddenly you've made yourself as the programmer here responsible to synchronize things like the SimVar changing from the outside into your local state.

Your life will be a lot easier, if you just use this state variable for the edit case, with something like this:

```suggestion
    const [newDigits, setNewDigits] = useState<null | number[]>(null);
```

`null` would indicate "no edit is in progress", and then further down you can just use the SimVar itself. In fact, you could put that right here, and we've basically already got the entire display logic done:

```suggestion
    const [newDigits, setNewDigits] = useState<null | number[]>(null);
    const [transponderCode, setTransponderCode] = useSplitSimVar('TRANSPONDER CODE:1', 'number', 'K:XPNDR_SET', 'Bco16', 250);
    const codeInDisplay: number[] = newDigits !== null // if we're in the process of editing...
      ? newDigits // ...just use the WIP digits, otherwise...
      : getDigitsFromCode(transponderCode) // ...just get the digits from the SimVar
```

Now you can think of `newDigits` as the "draft version". If it exists we show it, otherwise just use the known good version. We also don't need to worry about `currentEnteredDigits` in state any more - we can just figure them out by looking at newDigits being null or not.

You'll see below how this simplifies your life! 😃